MHVLib/Clients/MHVVocabularyClient.m

```diff
@@ -0,0 +1,249 @@
+}
+
+- (void)getVocabulariesWithVocabularyKeys:(MHVVocabularyKeyCollection *)vocabularyKeys
+                           cultureIsFixed:(BOOL)cultureIsFixed
+                               completion:(void(^)(MHVVocabularyCodeSetCollection* _Nullable vocabularies, NSError *_Nullable error))completion
+{
+    MHVASSERT_PARAMETER(vocabularyKeys);
+    MHVASSERT_PARAMETER(completion);
+    
+    if (!completion || !vocabularyKeys)
+    {
+        return;
+    }
+    
+    XWriter *writer = [[XWriter alloc] initWithBufferSize:2048];
+    [writer writeStartElement:@"info"];
+    [writer writeStartElement:@"vocabulary-parameters"];
+    
+    for (MHVVocabularyKey *key in vocabularyKeys)
+    {
+        [writer writeStartElement:@"vocabulary-key"];
+        [key serialize:writer];
+        [writer writeEndElement];
+    }
+    
+    [writer writeElement:@"fixed-culture" boolValue:cultureIsFixed];
+    [writer writeEndElement];   // </vocabulary-parameters>
+    [writer writeEndElement];   // </info>
+    
+    MHVMethod *method = [MHVMethod getVocabulary];
+    method.parameters = [writer newXmlString];
+    [self.connection executeHttpServiceOperation:method completion:^(MHVServiceResponse * _Nullable response, NSError * _Nullable error)
+    {
+        if (error)
+        {
+            completion(nil, error);
+            return;
+        }
+        
+        MHVVocabularyCodeSetCollection *vocabularies = (MHVVocabularyCodeSetCollection*)[XSerializer newFromString:response.infoXml withRoot:@"info" andElementName:@"vocabulary" asClass:[MHVVocabularyCodeSet class] andArrayClass:[MHVVocabularyCodeSetCollection class]];
+        
+        completion(vocabularies, nil);
+        return;
+    }];
+}
+
+- (void)searchVocabularyKeysWithSearchValue:(NSString *)searchValue
+                                 searchMode:(MHVSearchMode)searchMode
+                                 maxResults:(NSNumber * _Nullable)maxResults
+                                 completion:(void(^)(MHVVocabularyKeyCollection * _Nullable vocabularyKeys, NSError * _Nullable))completion
+{
+    MHVASSERT_PARAMETER(searchValue);
+    MHVASSERT_PARAMETER(searchMode);
+    MHVASSERT_PARAMETER(completion);
+    
+    if (!searchValue || !searchMode || !completion)
+    {
+        return;
+    }
+    
+    MHVMethod * method = [self getVocabularySearchMethodWithSearchValue:searchValue andSearchMode:searchMode andMaxResults:maxResults andVocabularyKey:nil];
+    
+    [self.connection executeHttpServiceOperation:method completion:^(MHVServiceResponse * _Nullable response, NSError * _Nullable error)
+     {
+         if (error)
+         {
+             completion(nil, error);
+             return;
+         }
+         
+         MHVVocabularyKeyCollection *vocabularyKeys = (MHVVocabularyKeyCollection*)[XSerializer newFromString:response.infoXml withRoot:@"info" andElementName:@"vocabulary-key" asClass:[MHVVocabularyKey class] andArrayClass:[MHVVocabularyKeyCollection class]];
+         
```

Comment:
SearchVocabulary actually has two possible return types currently. Depending on the parameters provided to the call, two different things happen. If you provide a VocabularyKey, then the CodeItems for that key are searched and the result is a VocabularyCodeSet. If you do not provide a VocabularyKey then the call searches all VocabularyKeys and the result is VocabularyKeyInfo. I have reached out to the coreplat team because this is poor API design and we should have two distinct methods instead of a single method with variable functionality. see https://developer.healthvault.com/Methods/ResponseSchema?Name=SearchVocabulary&Version=1